### ISSUE TITLE
webpack 2.2.0-rc.2: "TypeError: Cannot read property 'size' of undefined"

### ISSUE THREAD

**Issue by jdb8** [numical/style-ext-html-webpack-plugin#11]
On the latest webpack RC (v2.2.0-rc.2), using the plugin produces the following traceback:

```
/Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/webpack/lib/performance/SizeLimitsPlugin.js:37
					return compilation.assets[file].size()
					                               ^

TypeError: Cannot read property 'size' of undefined
    at /Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/webpack/lib/performance/SizeLimitsPlugin.js:37:37
    at Array.map (native)
    at getEntrypointSize (/Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/webpack/lib/performance/SizeLimitsPlugin.js:36:6)
    at /Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/webpack/lib/performance/SizeLimitsPlugin.js:64:16
    at Array.forEach (native)
    at Compiler.<anonymous> (/Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/webpack/lib/performance/SizeLimitsPlugin.js:62:5)
    at Compiler.applyPluginsAsyncSeries1 (/Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/tapable/lib/Tapable.js:122:13)
    at Compiler.afterEmit (/Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/webpack/lib/Compiler.js:340:8)
    at Compiler.<anonymous> (/Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/webpack/lib/Compiler.js:335:14)
    at /Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/async/dist/async.js:356:16
    at iteratorCallback (/Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/async/dist/async.js:936:13)
    at /Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/async/dist/async.js:840:16
    at /Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/graceful-fs/graceful-fs.js:43:10
    at /Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/graceful-fs/graceful-fs.js:43:10
    at /Users/jb/code/style-ext-html-webpack-plugin/dynavers_modules/webpack-2.2.0-rc.2/node_modules/graceful-fs/graceful-fs.js:43:10
    at FSReqWrap.oncomplete (fs.js:123:15)
```

(this was produced from upgrading the version of webpack 2 used during test and running `npm run test:webpack2`). The same error occurs when using the plugin with a real toy project.

A quick glance appears to show the cause is the line deleting the extracted css file: https://github.com/numical/style-ext-html-webpack-plugin/blob/3e17a2517df6b733e13942b415870ca24adc735b/lib/plugin.js#L107 - apparently the (new?) SizeLimitsPlugin is expecting it to exist: https://github.com/webpack/webpack/blob/8bc25f65e60001861a2c9f9962ddfbe3a5aa9025/lib/performance/SizeLimitsPlugin.js#L37. Removing the deletion line appears to work, although the tests fail since they're currently testing that the file is deleted.

I'm not sure whether this is more of a bug with webpack, or if the recommendation is for plugins to avoid deleting emitted assets created by other plugins? At this point my knowledge of webpack plugin execution order and interop is limited, so please let me know if I should create an issue on webpack itself.

**Comment by numical** [comment]
Thanks for the comprehensive issue report - **ON XMAS EVE**  :-p
I will have to look into what changes the RC release has introduced *sigh*.
I really wish the javascript community had not gone down the route of always using the shiniest beta versions of everything - unstable API's, especially unstable, unspec'd API's create so much unnecessary work.
Anyway enough wingeing: I always suspected the file delete - until recently it was covered by a conditional flag.  Naturally as soon as I took the flag away, it looks like it might now be needed.
As to your question about plugin interop - it's probably about time I learnt about that too!  
Leave it with me - and Happy New Year.

**Comment by jdb8** [comment]
@numical: no problem, thanks for taking a look.

I did a little more investigation and I'm leaning towards the opinion that this should be fixed on the webpack side. I can't find anything that officially discourages deletion of compilation assets during/after emit (although trying to google this stuff is basically impossible), and the code in the SizeLimitsPlugin seems easily refactor-able to simply ignore anything that's no longer in the set of assets. Additionally, a quick search seems to show that it's not just you doing it this way: https://github.com/search?q=%22delete+compilation.assets%22&ref=searchresults&type=Code&utf8=%E2%9C%93 - so any of these plugins would also be affected.

I'll likely create an issue/PR on webpack when I get around to it and see what they think about this. It's possible there's an alternative supported method for removing emitted assets. I tried briefly asking in the gitter but didn't get a response so far.

As for RC/betas breaking stuff, I feel your pain. But it's at least nice to find this stuff before v2 is released for real I guess ;)

**Comment by numical** [comment]
Hold off on any issue/PR for now.
The issue is caused as there is more than one reference to the generated file within the `compilation` (something I knew but thought I could get away with as I was operating so late in the compilation lifecycle).  If I am deleting a file, it really is incumbent on myself to clear all the references to that file (... though a simpler API ... etc. ... grumble ...).  
Gimme a chance (aka 2-3 days) to do a little debug digging before we bother the webpack masters...

**Comment by numical** [comment]
OK, looks like all I needed was to remove the file from its parent `chunk` as well as the assets.
I cannot think of any gotcha this would introduce (not at the `emit` stage, anyhow) so I declare this issue fixed at release 3.0.7.
Let me know if you find otherwise...

**Comment by jdb8** [comment]
Nice! FYI I filed https://github.com/webpack/webpack/issues/3596 and it's already been fixed, so I think you should probably be good to revert that (if you want) once the next version comes out.
